crates/bevy_reflect/src/utility.rs

```diff
@@ -110,35 +157,6 @@ impl NonGenericTypeInfoCell {
 /// #   fn clone_value(&self) -> Box<dyn Reflect> { todo!() }
 /// # }
 /// ```
-pub struct GenericTypeInfoCell(OnceBox<RwLock<HashMap<TypeId, &'static TypeInfo>>>);
+pub type GenericTypeInfoCell = GenericDataCell<TypeInfo>;
 
-impl GenericTypeInfoCell {
-    /// Initialize a [`GenericTypeInfoCell`] for generic types.
-    pub const fn new() -> Self {
-        Self(OnceBox::new())
-    }
-
-    /// Returns a reference to the [`TypeInfo`] stored in the cell.
-    ///
-    /// This method will then return the correct [`TypeInfo`] reference for the given type `T`.
-    /// If there is no [`TypeInfo`] found, a new one will be generated from the given function.
-    pub fn get_or_insert<T, F>(&self, f: F) -> &TypeInfo
-    where
-        T: Any + ?Sized,
-        F: FnOnce() -> TypeInfo,
-    {
-        let type_id = TypeId::of::<T>();
-        let mapping = self.0.get_or_init(|| Box::new(RwLock::default()));
-        if let Some(info) = mapping.read().get(&type_id) {
-            return info;
-        }
-
-        mapping.write().entry(type_id).or_insert_with(|| {
-            // We leak here in order to obtain a `&'static` reference.
-            // Otherwise, we won't be able to return a reference due to the `RwLock`.
-            // This should be okay, though, since we expect it to remain statically
-            // available over the course of the application.
-            Box::leak(Box::new(f()))
-        })
-    }
-}
+pub type GenericTypeNameCell = GenericDataCell<String>;
```

Comment:
Could you provide some documentation here? Especially for those who need to implement `TypeName` manually (for whatever reason).